File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    static boolean[] selected;
    static int answer;
    public int solution(String begin, String target, String[] words) {
        answer = Integer.MAX_VALUE;
        
        selected = new boolean[words.length];
        boolean check = false;
        for(int i=0; i<words.length; i++){
            if(words[i].equals(target)){
                check = true;
                break;
            }
        }
        
        if(check) dfs(0, 0, begin, target, words);
        else answer = 0;
        
        return answer;
    }
    static void dfs(int idx, int cnt, String cur, String target, String[] words){
        if(cur.equals(target)){
            answer = Math.min(answer, cnt);
            return;
        }
        
        if(cnt >= answer) return;
        
        for(int i=0; i<words.length; i++){
            if(selected[i]) continue;
            
            int n = 0;
            for(int j=0; j<cur.length(); j++){
                if(cur.charAt(j) == words[i].charAt(j)) n++;
            }
            if(n == cur.length() - 1){
                selected[i] = true;
                dfs(idx + 1, cnt + 1, words[i], target, words);
                selected[i] = false;
            }
        }
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    
    answer = bfs(begin, target, words)
    return answer

def bfs(begin, target, words):
    s = set()
    q = deque()
    q.append((begin, 0))
    
    cnt = 0
    while q:
        nowWord, nowCnt = q.popleft()
        
        if nowWord == target:
            return nowCnt
        
        for word in words:
            changeCnt = 0
            
            for i in range(len(word)):
                if word[i:i+1] != nowWord[i:i+1]:
                    changeCnt += 1
            
            if changeCnt == 1 and word not in s:
                s.add(word)
                q.append((word, nowCnt + 1))
    return 0        
            